### PR TITLE
common/tracer: Tracer implementation using opentelemetry sdk

### PR DESCRIPTION
**Pull request by zenomri**
common/tracer: Tracer implementation using opentelemetry sdk

this is the implementation of the Tracer class after moving to use [opentelemetry SDK](https://github.com/open-telemetry/opentelemetry-cpp)
the first commit is the previous PR (#42878) changes (in which we were using jaeger SDK)
and the second one is the changes due to moving to opentelemetry SDK.

this PR is a draft until we will have opentelemtry library compiled with ceph build.

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/common/tracer.cc [hunk 1]
```diff
@@ -1,86 +1,60 @@
...
+  if (tracer) {
+    tracer->CloseWithMicroseconds(1);
+  }
+}
 
-  void finish_span(const jspan& span) {
-    if (span) {
-      span->Finish();
-    }
+opentelemetry::nostd::shared_ptr<opentelemetry::trace::Span> Tracer::start_trace(opentelemetry::nostd::string_view trace_name) {
```
Comment (yuvalif) [inline src/common/tracer.cc:35]: is there an api that does not use a shared pointer?

**Comment by zenomri** [comment]
jenkins test make check

**Comment by yuvalif** [comment]
we decided that this PR would be merged with the code compiled out until we run performance tests against the feature.
in this case, we can go on, build and run teuthology against without waiting for: https://github.com/ceph/ceph/pull/43598

**Comment by zenomri** [comment]
jenkins test make check arm64

**Comment by zenomri** [comment]
jenkins test make check arm64

**Comment by yuvalif** [comment]
@zenomri this PR is ready for teuthology. please take out of "draft" state and run teuthology against it when feature is compiled out.

**Comment by zenomri** [comment]
> @zenomri this PR is ready for teuthology. please take out of "draft" state and run teuthology against it when feature is compiled out.

I have already run rgw test suite 
here is the results
http://pulpito.front.sepia.ceph.com/ozeneva-2021-11-03_12:43:24-rgw-wip-omri-tracer-opentelemetry-distro-basic-smithi

rados test suite is still running

**Comment by yuvalif** [comment]
> > @zenomri this PR is ready for teuthology. please take out of "draft" state and run teuthology against it when feature is compiled out.
> 
> I have already ran rgw test suite here is the results http://pulpito.front.sepia.ceph.com/ozeneva-2021-11-03_12:43:24-rgw-wip-omri-tracer-opentelemetry-distro-basic-smithi
> 
> rados test suite is still running

RGW results looks good. out of 8 failures we have:
* 5 "health warnings"
* 1 sts test failure (known issue, exists in master baseline)
* 1 multisite failure (known issue, exists in master baseline)
* one is python issue with the test code: `ERROR: Cannot install PrettyTable<0.8 and >=0.7.1 because these package versions have conflicting dependencies.`

lets wait for OSD results, and compare them to their master baseline.

**Comment by zenomri** [comment]
rados test suite results:
http://pulpito.front.sepia.ceph.com/ozeneva-2021-11-04_23:53:22-rados-wip-omri-tracer-opentelemetry-distro-basic-smithi/

looks good in comparison to [master results](http://pulpito.front.sepia.ceph.com/yuriw-2021-10-29_23:44:51-rados-wip-yuri-master-10.29.21-distro-basic-smithi/)

**Review by yuvalif — APPROVED** [review]

**Comment by ideepika** [comment]
> rados test suite results: http://pulpito.front.sepia.ceph.com/ozeneva-2021-11-04_23:53:22-rados-wip-omri-tracer-opentelemetry-distro-basic-smithi/
> 
> looks good in comparison to [master results](http://pulpito.front.sepia.ceph.com/yuriw-2021-10-29_23:44:51-rados-wip-yuri-master-10.29.21-distro-basic-smithi/)
> 
> @ideepika could you please look at the results?

will review today, thanks!

**Comment by ideepika** [comment]
@zenomri I think we need to investigate some more, for example having a core dump means a definite crash, which I don't see in master baseline: 
http://qa-proxy.ceph.com/teuthology/ozeneva-2021-11-04_23:53:22-rados-wip-omri-tracer-opentelemetry-distro-basic-smithi/6486035/teuthology.log 
job timeout seems related to this PR as well http://qa-proxy.ceph.com/teuthology/ozeneva-2021-11-04_23:53:22-rados-wip-omri-tracer-opentelemetry-distro-basic-smithi/6485783/teuthology.log 
I would suggest to go to the /a/ozeneva-2021-11-04_23:53:22-rados-wip-omri-tracer-opentelemetry-distro-basic-smithi/6486035/remote/smithi142/coredump and try to see if the issue is related. 
you can use teuthology rerun with filter to investigate individual issues if testing fix https://docs.ceph.com/en/pacific/dev/developer_guide/running-tests-using-teuth/

**Review by ideepika — DISMISSED** [review]
@zenomri I think we need to investigate some more, for example having a core dump means a definite crash, which I don't see in master baseline:
http://qa-proxy.ceph.com/teuthology/ozeneva-2021-11-04_23:53:22-rados-wip-omri-tracer-opentelemetry-distro-basic-smithi/6486035/teuthology.log
job timeout seems related to this PR as well http://qa-proxy.ceph.com/teuthology/ozeneva-2021-11-04_23:53:22-rados-wip-omri-tracer-opentelemetry-distro-basic-smithi/6485783/teuthology.log
I would suggest to go to the /a/ozeneva-2021-11-04_23:53:22-rados-wip-omri-tracer-opentelemetry-distro-basic-smithi/6486035/remote/smithi142/coredump and try to see if the issue is related.
you can use teuthology rerun with filter to investigate individual issues if testing fix https://docs.ceph.com/en/pacific/dev/developer_guide/running-tests-using-teuth/

**Comment by zenomri** [comment]
> @zenomri I think we need to investigate some more, for example having a core dump means a definite crash, which I don't see in master baseline: http://qa-proxy.ceph.com/teuthology/ozeneva-2021-11-04_23:53:22-rados-wip-omri-tracer-opentelemetry-distro-basic-smithi/6486035/teuthology.log job timeout seems related to this PR as well http://qa-proxy.ceph.com/teuthology/ozeneva-2021-11-04_23:53:22-rados-wip-omri-tracer-opentelemetry-distro-basic-smithi/6485783/teuthology.log I would suggest to go to the /a/ozeneva-2021-11-04_23:53:22-rados-wip-omri-tracer-opentelemetry-distro-basic-smithi/6486035/remote/smithi142/coredump and try to see if the issue is related. you can use teuthology rerun with filter to investigate individual issues if testing fix https://docs.ceph.com/en/pacific/dev/developer_guide/running-tests-using-teuth/

thanks! I will investigate it.

**Comment by zenomri** [comment]
I reran the failed/dead tests. the results are much better
http://pulpito.front.sepia.ceph.com/ozeneva-2021-11-08_20:31:46-rados-wip-omri-tracer-opentelemetry-distro-basic-smithi/

the 4 failed tests are also failed in the master baseline,
and those dead tests are also dead in other teuthology jobs, @ideepika  I haven't saw anything related to this PR, do you think it is related? or it's a known problem?

**Comment by ideepika** [comment]
> the 4 failed tests are also failed in the master baseline, and those dead tests are also dead in other teuthology jobs, @ideepika I haven't saw anything related to this PR, do you think it is related? or it's a known problem?

great news! can you share the other jobs having those test failures, or perhaps find a tracker ticket https://tracker.ceph.com/ if it does not exists can you create one and share? Thanks

**Comment by zenomri** [comment]
> > the 4 failed tests are also failed in the master baseline, and those dead tests are also dead in other teuthology jobs, @ideepika I haven't saw anything related to this PR, do you think it is related? or it's a known problem?
> 
> great news! can you share the other jobs having those test failures, or perhaps find a tracker ticket https://tracker.ceph.com/ if it does not exists can you create one and share? Thanks

http://pulpito.front.sepia.ceph.com/yuriw-2021-11-03_18:53:10-rados-wip-yuri5-testing-2021-11-02-0745-distro-basic-smithi/
http://pulpito.front.sepia.ceph.com/yuriw-2021-11-07_14:27:28-rados-wip-yuri-testing-master-11.5.21-distro-basic-smithi/

**Review by ideepika — APPROVED** [review]

**Pull request merged**